src/ng/compile.js

```diff
@@ -2419,11 +2421,18 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
           }
         });
 
-        // Trigger the `$onInit` method on all controllers that have one
+        // Trigger the `$onInit` method on all controllers that have one,
+        // and trigger `$onDestroy` method if present and when the element emits `$destroy` event
         forEach(elementControllers, function(controller) {
           if (isFunction(controller.instance.$onInit)) {
             controller.instance.$onInit();
           }
+
+          $element.on('$destroy', function() {
```

Comment:
The `$destroy` event won't function properly when you have a `transclude: "element"` directive since events are not fired on the comment node that becomes `$element`.

http://jsfiddle.net/rogqp47y/17/